### ISSUE TITLE
Corutines not working with signal spider_opened - spider freezes

### ISSUE THREAD

**Issue by ivanprado** [scrapy/scrapy#4893]
### Description

The following spider was run with `asyncio` reactor enabled. It freezes:

```py
from scrapy import Spider, signals
from scrapy.http import Response


class AsyncioSpider(Spider):
    name = "asyncio"
    start_urls = ["http://blog.scrapinghub.com"]

    @classmethod
    def from_crawler(cls, crawler, *args, **kwargs):
        spider = super(AsyncioSpider, cls).from_crawler(crawler, *args, **kwargs)
        crawler.signals.connect(spider.spider_opened, signals.spider_opened)
        return spider

    async def spider_opened(self):
        print("hey!")

    def parse(self, response: Response):
        yield dict(url=response.url, lenght=len(response.text))
```

When Ctrl+C is pressed, the following exception can be seen:

```
2020-11-18 10:38:09 [twisted] CRITICAL: 
Traceback (most recent call last):
  File "/home/ivan/Documentos/scrapinghub/dev/autocrawl_env/lib/python3.8/site-packages/scrapy/crawler.py", line 109, in stop
    yield defer.maybeDeferred(self.engine.stop)
  File "/home/ivan/Documentos/scrapinghub/dev/autocrawl_env/lib/python3.8/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/ivan/Documentos/scrapinghub/dev/autocrawl_env/lib/python3.8/site-packages/scrapy/core/engine.py", line 87, in stop
    raise RuntimeError("Engine not running")
RuntimeError: Engine not running
```

Note that if `async` is removed from the method `spider_opened`, then it works. It suggests that this signal is not supporting coroutines properly. 

### Steps to Reproduce

Run the given spider with asyncio reactor enabled.

**Expected behavior:** [What you expect to happen]

The spider to run properly and corotines working well with `spider_opened` signal.

**Actual behavior:** [What actually happens]

The spider freezes

**Reproduces how often:** [What percentage of the time does it reproduce?]

Always

### Versions

```
Scrapy       : 2.3.0
lxml         : 4.5.2.0
libxml2      : 2.9.10
cssselect    : 1.1.0
parsel       : 1.6.0
w3lib        : 1.22.0
Twisted      : 20.3.0
Python       : 3.8.5 (default, Jul 20 2020, 19:48:14) - [GCC 7.5.0]
pyOpenSSL    : 19.1.0 (OpenSSL 1.1.1g  21 Apr 2020)
cryptography : 3.1
Platform     : Linux-5.7.1-050701-generic-x86_64-with-glibc2.27
```

**Comment by Gallaecio** [comment]
https://github.com/scrapy/scrapy/pull/4872

**Comment by elacuesta** [comment]
@ivanprado Out of curiosity, is the reported `2.3.0` version correct? I cannot reproduce with 1278e76d9093b1c5c9ec810768d1066772a8a134. It is indeed a problem with `2.4.0` (c340e72988fc6ec615b7b9851c3d28c16c26a839), but it was fixed by #4872 (and released as `2.4.1`- 26836c4e1ae9588ee173c5977fc6611364ca7cc7) as @Gallaecio said.

**Comment by ivanprado** [comment]
@elacuesta I can confirm this fails in Scrapy 2.4.0 but does not fail either for 2.3.0 or 2.4.1. So everything is right :+1:
